src/compiler/checker.ts

```diff
@@ -12776,17 +12776,16 @@ namespace ts {
                 if (isFreshLiteralType(target)) {
                     target = (<FreshableType>target).regularType;
                 }
-                if (source.flags & TypeFlags.Substitution) {
-                    source = (<SubstitutionType>source).substitute;
-                }
-                if (target.flags & TypeFlags.Substitution) {
-                    target = (<SubstitutionType>target).typeVariable;
-                }
-                if (source.flags & TypeFlags.Simplifiable) {
-                    source = getSimplifiedType(source, /*writing*/ false);
-                }
-                if (target.flags & TypeFlags.Simplifiable) {
-                    target = getSimplifiedType(target, /*writing*/ true);
+                while (true) {
+                    const s = source.flags & TypeFlags.Substitution ? source = (<SubstitutionType>source).substitute :

```

Comment:
Would it ever be worth updating `substitute` in-place to collapse multiple links down to one?

Reply:
I don't think we can - the associated `typeVariable` of nested `SubstitutionType`s can (and most often does) differ. ie, a `T` substituted with `T & U` where `U` is substituted with `U & string` - while here we want to fully flatten, in other situations I think we want to keep the knowledge of what variables expanded to what.

Reply:
Right; I forgot that these are more like narrowings rather than unification substitutions.